tokio-trace/src/level_filters.rs

```diff
@@ -0,0 +1,121 @@
+use std::cmp::Ordering;
+use tokio_trace_core::Level;
+
+/// Describes the level of verbosity of a span or event.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct LevelFilter(LevelFilterInner, Option<Level>);
+
+impl LevelFilter {
+    /// The "zero" level.
+    ///
+    /// Designates no logging.
+    pub const ZERO: LevelFilter = LevelFilter(LevelFilterInner::Zero, None);
+    /// The "error" level.
+    ///
+    /// Designates very serious errors.
+    pub const ERROR: LevelFilter = LevelFilter(LevelFilterInner::Error, Some(Level::ERROR));
+    /// The "warn" level.
+    ///
+    /// Designates hazardous situations.
+    pub const WARN: LevelFilter = LevelFilter(LevelFilterInner::Warn, Some(Level::WARN));
+    /// The "info" level.
+    ///
+    /// Designates useful information.
+    pub const INFO: LevelFilter = LevelFilter(LevelFilterInner::Info, Some(Level::INFO));
+    /// The "debug" level.
+    ///
+    /// Designates lower priority information.
+    pub const DEBUG: LevelFilter = LevelFilter(LevelFilterInner::Debug, Some(Level::DEBUG));
+    /// The "trace" level.
+    ///
+    /// Designates very low priority, often extremely verbose, information.
+    pub const TRACE: LevelFilter = LevelFilter(LevelFilterInner::Trace, Some(Level::TRACE));
+}
+
+#[repr(usize)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
+enum LevelFilterInner {
+    Zero = 0,
+    /// The "error" level.
+    ///
+    /// Designates very serious errors.
+    Error,
+    /// The "warn" level.
+    ///
+    /// Designates hazardous situations.
+    Warn,
+    /// The "info" level.
+    ///
+    /// Designates useful information.
+    Info,
+    /// The "debug" level.
+    ///
+    /// Designates lower priority information.
+    Debug,
+    /// The "trace" level.
+    ///
+    /// Designates very low priority, often extremely verbose, information.
+    Trace,
+}
+
+impl PartialEq<LevelFilter> for Level {
```

Comment:
I think the reason these traits had to be implemented for `Level` rather than for `LevelFilter` is because the `Level` is on the right-hand side of the comparison and the `LevelFilter` is on the left. We could reverse the comparison if we wanted the impl to be `PartialEq<Level> for LevelFilter` instead, like this:
```rust
if $crate::level_filters::STATIC_MAX_LEVEL > $lvl
```

However, this seems to work, so I'm fine either way.

Reply:
As a side note, I'm actually surprised the compiler accepts this due to the [orphan rules](http://smallcultfollowing.com/babysteps/blog/2015/01/14/little-orphan-impls/); neither `PartialEq` nor `Level` are defined in this crate. I suppose since the `PartialEq` (and `PartialOrd`) traits are being implemented with a _type parameter_ defined in this crate (`LevelFilter`) this is considered acceptable.

I've just never seen this before --- it's possible the compiler only recently started accepting this?

Reply:
This must be something new, I can `impl` for almost all kind of things :/